lib/cog/nested_file.ex

```diff
@@ -0,0 +1,86 @@
+defmodule Cog.NestedFile do
+  @moduledoc """
+  Stores an arbitrary String on the filesystem of the Cog host in a file
+  named by combining the provided key and extension. Files are written to
+  a directory hierarchy that is created by splitting the filename into
+  two character segments and joining up to three of those segments with a
+  provided list of base path segments.
+
+  Note: The key is sanitized to remove dangerous characters, but the
+  base paths and extension are not. Make sure not to use user entered
+  values for these unsafe arguments.
+
+  Example: If the replace function was called with the following options:
+
+  base_paths = [ "commands", "tee" ]
+  key = "myfilename/../foo"
+  ext = "json"
+
+  The following would be created on disk:
+
+  .
+  └── commands
+      └── tee
+          └── my
+              └── fi
+                  └── le
+                      └── myfilenamefoo.json
+
+  This directory structure is created in order to deal with the fact that
+  some filesystems demonstrate poor performance characteristics when working
+  with directories that contain a very large number of files.
+  """
+  require Logger
+
+  def fetch(base_paths, key, ext \\ "data") do
```

Comment:
It was mostly there because I didn't know if we'd want to use `NestedFile` in other places in Cog since it's fairly generic.